File: Divide_and_Conquer/493.Reverse_Pairs/493.Reverse_pairs.py
class Solution:
    def reversePairs(self, nums: 'List[int]') -> 'int':
        return self.mergesort(nums, 0, len(nums) - 1)
    def mergesort(self, nums:'List[int]', left, right) -> 'int':
        if left >= right:
            return 0
        mid = left + (right - left) // 2
        res = self.mergesort(nums, left, mid) + self.mergesort(nums, mid + 1, right)
        
        j = mid + 1
        for i in range(left, mid + 1):
            while j <= right and nums[i] > 2 * nums[j]:
                j += 1
            res += j - (mid + 1)
        nums[left: right + 1] = sorted(nums[left: right + 1])
        '''
        L, R = nums[left: mid + 1], nums[mid + 1: right + 1]
        p, q = 0, 0
        for k in range(left, right + 1):
            if q == len(R) or (p < len(L) and L[p] < R[q]):
                nums[k] = L[p]
                p += 1
            else:
                nums[k] = R[q]
                q += 1
        '''
        return res

File: Divide_and_Conquer/493.Reverse_Pairs/493.Reverse_Pairs.cpp
class Solution {
public:
    int reversePairs(vector<int>& nums) {
        return mergesort(nums, 0, nums.size() - 1);
    }
    
    int mergesort(vector<int> &nums, int left, int right) {
        if (left >= right) return 0;
        int mid = left + (right - left) / 2;
        int res = mergesort(nums, left, mid) + mergesort(nums, mid + 1, right);
        for (int i = left, j = mid + 1; i <= mid; i++) {
            while (j <= right && (long)nums[i] > 2 * (long)nums[j]) j++;
            res += j - (mid + 1);
        }
        merge(nums, left, mid, right);
        return res;
    }
    
    void merge(vector<int> &nums, int left, int mid, int right) {
        int n1 = mid - left + 1, n2 = right - mid;
        int L[n1], R[n2];
        for (int i = left; i <= mid; i++)
            L[i - left] = nums[i];
        for (int i = mid + 1; i <= right; i++)
            R[i - (mid + 1)] = nums[i];
        int i = 0, j = 0;
        for (int k = left; k <= right; k++){
            if (j >= n2 || (i < n1 && L[i] < R[j])) nums[k] = L[i++];
            else nums[k] = R[j++];
        }
    }
};
